src/timely-util/src/builder_async.rs

```diff
@@ -524,6 +524,68 @@ impl<G: Scope> OperatorBuilder<G> {
         button
     }
 
+    /// Creates a fallible operator implementation from supplied logic constructor. If the `Future`
+    /// resolves to an error it will emitted in the returned error stream and then the operator
+    /// will wait indefinitely until the shutdown button is pressed.
+    ///
+    /// # Capability handling
+    ///
+    /// Unlike [`OperatorBuilder::build`], this method does not give owned capabilities to the
+    /// constructor. All initial capabilities are wrapped in an `Option` and a mutable reference to
+    /// them is given instead. This is done to avoid storing owned capabilities in the state of the
+    /// logic future which would make using the `?` operator unsafe, since the frontiers would
+    /// incorrectly advance, potentially causing incorrect actions downstream.
+    ///
+    /// ```ignore
+    /// builder.build_fallible(|caps| Box::pin(async move {
+    ///     // Assert that we have the number of capabilities we expect
+    ///     // `cap` will be a `&mut Option<Capability<T>>`:
+    ///     let [cap]: &mut [_; 1] = caps.try_into().unwrap();
+    ///
+    ///     // Using cap to send data:
+    ///     output.give(cap.as_ref().unwrap(), 42);
+    ///
+    ///     // Using cap to downgrade it:
+    ///     cap.as_mut().unwrap().downgrade();
+    ///
+    ///     // Explicitly dropping the capability:
+    ///     // Simply running `drop(cap)` will only drop the reference and not the capability itself!
+    ///     *cap = None;
+    ///
+    ///     // !! BIG WARNING !!:
+    ///     // It is tempting to `take` the capability out of the option for convenience. This will
+    ///     // move the capability into the future state, tying its lifetime to it, which will get
+    ///     // dropped when an error is hit, causing incorrect progress statements.
+    ///     let cap = cap.take().unwrap(); // DO NOT DO THIS
```

Comment:
Because the logic needs to be able to drop the capability. if you did `foo = None` on a `Option<&mut Cap>` it would just drop the mut ref, which is a no-op.

Reply:
ah yes!